### ISSUE TITLE
ebay.com: ads

### ISSUE THREAD

**Issue by sha-265** [uBlockOrigin/uAssets#23491]
### Prerequisites

- [X] This is **NOT** a [_YouTube_](https://www.github.com/uBlockOrigin/uAssets/issues/20586), [_Facebook_](https://www.github.com/uBlockOrigin/uAssets/issues/3367), [_Twitch_](https://www.github.com/uBlockOrigin/uAssets/issues/5184) or [_a shortener/hosting site_](https://github.com/uBlockOrigin/uAssets/discussions/17361) report. _These sites MUST be reported by clicking their **respective links**_.
- [X] I read and understand the [policy about what is a valid filter issue](https://github.com/uBlockOrigin/uAssets/blob/master/README.md#uassets).
- [X] I verified that this issue is not a duplicate. (Use this [button](https://user-images.githubusercontent.com/585534/146582579-c32707a0-36a1-4cc5-ad50-83172c9f67a8.png) to find out.)
- [X] I did not remove any of the [default filter lists](https://github.com/uBlockOrigin/uAssets/assets/66517106/6e8d024a-5869-47cf-b142-b2357eb6c7d0), or I have verified that the issue was not caused by removing any of the default lists.
- [X] I did not enable additional filter lists, or I have verified that the issue still occurs without enabling additional filter lists.
- [X] I do not have custom filters/rules, or I have verified that the issue still occurs without custom filters/rules.
- [X] I am not using uBlock Origin along with other content blockers.
- [X] I have verified that the web browser's built-in content blocker/tracking protection, [network wide/DNS blocking](https://en.wikipedia.org/wiki/DNS_blocking), or my VPN is not causing the issue.
- [X] I have verified that other extensions are not causing the issue.
- [X] If this is about a breakage or detection, I have verified that it is caused by uBlock Origin and isn't a site issue.
- [ ] I did not answer truthfully to **ALL** the above checkboxes.

### URL address of the web page

`https://www.ebay.com/sch/i.html`

### Category

ads

### Description

Some sponsored listings are appears in search results in eBay

### Other extensions used

CanvasBlocker

### Screenshot(s)

<details>
<summary>Screenshot(s)</summary>

![image](https://github.com/uBlockOrigin/uAssets/assets/4103710/fdf3de9f-0493-4084-a11f-271699c96968)

</details>

### Configuration

<details>

```yaml
uBlock Origin: 1.57.2
Firefox: 125
filterset (summary):
 network: 145565
 cosmetic: 47666
 scriptlet: 19277
 html: 1793
listset (total-discarded, last-updated):
 removed:
  plowe-0: null
 added:
  adguard-spyware-url: 1504-120, now
  block-lan: 66-0, now
  fanboy-thirdparty_social: 68-0, now
  dpollock-0: 11669-369, now
  ISR-0: 1866-3, now
 default:
  user-filters: 5-0, never
  ublock-filters: 37821-109, now
  ublock-badware: 8450-0, now
  ublock-privacy: 875-2, now
  ublock-unbreak: 2321-2, now
  ublock-quick-fixes: 183-18, now
  easylist: 86849-174, now
  easyprivacy: 50825-64, now
  urlhaus-1: 12750-0, now
filterset (user): [array of 6 redacted]
trustedset:
 added: [array of 1 redacted]
hostRuleset:
 added: [array of 3 redacted]
userSettings:
 advancedUserEnabled: true
 cloudStorageEnabled: true
 cnameUncloakEnabled: false
hiddenSettings: [none]
supportStats:
 allReadyAfter: 4510 ms (selfie)
 maxAssetCacheWait: 229 ms
 cacheBackend: indexedDB
popupPanel:
 blocked: 9
 network:
  ebay.com: 4
  ebaystatic.com: 5
 extended:
  ##div[title="ADVERTISEMENT"]
```
</details>

**Comment by stephenhawk8054** [comment]
They are using `content` inside `::before` so I don't think this can be addressed

**Comment by sha-265** [comment]
> They are using `content` inside `::before` so I don't think this can be solved.

can't we use `has()` element `<span data-w="Sponsored">`, or something like that?

**Comment by stephenhawk8054** [comment]
Every item has `<span data-w="Sponsored">`, it's just some show the text, and some hide the text

**Comment by gwarser** [comment]
span[data-w]:matches-css-before(content: /attr/)

**Comment by stephenhawk8054** [comment]
Oh, didn't know there's `:matches-css-before`, but looks like I'm mistaken, it might not be related to content

**Comment by sha-265** [comment]
so is it solvable or not?

**Comment by stephenhawk8054** [comment]
Others may find some ways, but currently I don't know how to address it.

It uses 2 `transform: rotate` steps to rotate the "sponsored" text. The sponsored items will be rotated back to normal position and we can see the "sponsored" text, others will be rotated out of view and we cannot see that text.

**Comment by gwarser** [comment]
> check the value of the ID in one element against the value of another attribute in listing items.

Should be possible with XPath.

**Comment by u-RraaLL** [comment]
I was wondering about that. But it's not within my skillset to attempt this.

    ##[aria-hidden="true"]>span.clipped[id]:not(:has(+[aria-labelledby]))
    ##.s-item [aria-labelledby]

**Comment by Yuki2718** [comment]
Are we talking about Sponsored in the screenshot? I don't see sponsored at all, maybe geolocation.

**Comment by mapx-** [comment]
see for example 
`https://www.ebay.com/b/32-Degrees/bn_7064048796`

**Comment by Yuki2718** [comment]
@u-RraaLL 
> The attribute value is mentioned in another element though, so we'd need a procedural that can check the value of the ID in one element against the value of another attribute in listing items.

What exactly?

**Comment by Yuki2718** [comment]
It seems Sponsored items have longer links like FB.

**Comment by u-RraaLL** [comment]
> What exactly?

I've mentioned them above. The value of id in `body>[aria-hidden="true"]>span.clipped[id]:not(:has(+[aria-labelledby]))` is the same as sponsored's `.s-item [aria-labelledby]`.

> It seems Sponsored items have longer links like FB. Or just `##.s-item__wrapper:has(a[href*="&amdata="])`?

Oh, it is? Guess we can check that.

That selector doesn't work for me though - it only finds one thing which is not part of the results?

https://imgur.com/RcLh4bB.png and https://imgur.com/TAZs58j.png

**Comment by Yuki2718** [comment]
Maybe geolocation, it's working for me and I'll add to EL unless it causes FP on your side.

**Comment by sha-265** [comment]
> Maybe geolocation, it's working for me and I'll add to EL unless it causes FP on your side. Whoops, the rule leaves leftover. So `##.s-item:has(a[href*="&amdata="])`

@Yuki2718, this filter is not working for me, sponsored items are still exist

**Comment by Yuki2718** [comment]
I know, but I first have to reproduce the ads you guys see. Which country do you access from?

**Comment by u-RraaLL** [comment]
You're right. The filter works fine when browsing by category. It just doesn't affect search results

    https://www.ebay.com/sch/i.html?_from=R40&_nkw=ps5&_sacat=0

**Comment by sha-265** [comment]
> I know, but I first have to reproduce the ads you guys see. Which country do you access from?

@Yuki2718, Israel

**Comment by u-RraaLL** [comment]
I was replying to mapx, who was referring to his URL which is "browse by a category".

I said it doesn't work on search results and provided an example for that.

**Comment by Yuki2718** [comment]
Test
```
ebay.com##:xpath(//li[contains(@class, "s-item ")]//span[@aria-labelledby=ancestor::body//span[@class="clipped" and @id and not(following-sibling::span[1 and @aria-labelledby])]/@id]/ancestor::li[contains(@class, "s-item ")])
```

**Comment by sha-265** [comment]
When creating the filter, it is working, but after page refresh the sponsored items are shown again

**Comment by Yuki2718** [comment]
The rule works with AdGuard but with noticeble delay to display the page. So I suspect it's a matter of `:watch-attr()`.

**Comment by Yuki2718** [comment]
Maybe `:watch-attr()`doesn't watch attributes specified in `:xpath()`? @gorhill

**Comment by gwarser** [comment]
Do not use wildcard, this executes xpath on every element (freezes my browser). Start with body or html.

**Comment by Yuki2718** [comment]
Should `body` be added to all existing `##:xpath` rules?

**Comment by u-RraaLL** [comment]
> No idea why it works fine in picker, but does not from my filters.

Maybe xpath executes too early?

    body>[aria-hidden="true"]>span.clipped[id]:not(:has(+[aria-labelledby]))

This selector works only after the page finished loading. Earlier there's another element there too. It corresponds to the "dummy sponsored" and gets deleted by an inline script.

**Comment by sha-265** [comment]
any updates?

**Comment by JobcenterTycoon** [comment]
Userscript:

```
ebay.com##+js(rpnt, script, removeItem("SRP.PLBTimestamp");o(true)}}}\,3e3)})}})(), removeItem("SRP.PLBTimestamp");o(true)}}}\,3e3)})}})(); var items=document.querySelectorAll(".srp-results > .s-item");for(let i=0;i<items.length;i++){var e=items[i].querySelector('span[aria-hidden="true"][data-w="Sponsored"]')\,t=(e=window.getComputedStyle(e).getPropertyValue("transform")).split("(")[1].split(")")[0].split("\,")\,r=t[0]\,s=t[1]\,a=t[2]\,l=t[3]\,o=Math.sqrt(r*r+s*s)\,p=s/o\,d=Math.round(Math.atan2(s\,r)*(180/Math.PI))\,n=items[i].querySelector('span[aria-hidden="true"][data-w="Sponsored"]')\,u=(n=window.getComputedStyle(n\,"::before").getPropertyValue("transform")).split("(")[1].split(")")[0].split("\,")\,m=u[0]\,y=u[1]\,_=u[2]\,g=u[3]\,S=Math.sqrt(m*m+y*y)\,$=y/S\,q=Math.round(Math.atan2(y\,m)*(180/Math.PI));d+q>=-1&&d+q<=1&&(console.log(d+q)\,items[i].style.display="none")})
```

**Comment by JobcenterTycoon** [comment]
Userscript which hide the ads in the search results:

```
ebay.com##+js(rpnt, script, removeItem("SRP.PLBTimestamp");o(true)}}}\,3e3)})}})(), removeItem("SRP.PLBTimestamp");o(true)}}}\,3e3)})}})(); !function(){let e=document.querySelectorAll(".srp-results > .s-item");for(let t=0;t<e.length;t++){let r=e[t].querySelector('span[aria-hidden="true"][data-w="Sponsored"]')\,l=(r=window.getComputedStyle(r).getPropertyValue("transform")).split("(")[1].split(")")[0].split("\,")\,o=l[0]\,a=l[1];l[2]\,l[3];let s=Math.round(Math.atan2(a\,o)*(180/Math.PI))\,n=e[t].querySelector('span[aria-hidden="true"][data-w="Sponsored"]')\,p=(n=window.getComputedStyle(n\,"::before").getPropertyValue("transform")).split("(")[1].split(")")[0].split("\,")\,d=p[0]\,u=p[1];p[2]\,p[3];let i=Math.round(Math.atan2(u\,d)*(180/Math.PI));s+i>=-1&&s+i<=1&&(e[t].style.display="none")}}();)
```

**Comment by Yuki2718** [comment]
So this can be fixed with rpnt. I still wonder why the xpath filter doesn't work with uBO.

**Comment by JobcenterTycoon** [comment]
So

```
ebay.*##+js(rpnt, script, removeItem("SRP.PLBTimestamp");o(true)}}}\,3e3)})}})(), removeItem("SRP.PLBTimestamp");o(true)}}}\,3e3)})}})(); !function(){let e=document.querySelectorAll(".srp-results > .s-item");for(let t=0;t<e.length;t++){let r=e[t].querySelector('span[aria-hidden="true"][data-w]')\,l=(r=window.getComputedStyle(r).getPropertyValue("transform")).split("(")[1].split(")")[0].split("\,")\,o=l[0]\,a=l[1];l[2]\,l[3];let s=Math.round(Math.atan2(a\,o)*(180/Math.PI))\,n=e[t].querySelector('span[aria-hidden="true"][data-w]')\,p=(n=window.getComputedStyle(n\,"::before").getPropertyValue("transform")).split("(")[1].split(")")[0].split("\,")\,d=p[0]\,u=p[1];p[2]\,p[3];let i=Math.round(Math.atan2(u\,d)*(180/Math.PI));s+i>=-1&&s+i<=1&&(e[t].style.display="none")}}();)
```

to cover also regional sites like `ebay.de` . For the xpath issue i don’t know i never work with xpath.

**Comment by loadstring1** [comment]
@JobcenterTycoon you forgot to close this issue

**Comment by stephenhawk8054** [comment]
There were no feedbacks at that time, so it's normal to close. If you see the ads, just continue here, no need for asking why closing the issue.

**Comment by sha-265** [comment]
I will ask why this issue is closed, because I can't reopen it myself. 

The problem never solved. No one asked for feedback.

**Comment by stephenhawk8054** [comment]
No one was able to reproduce the issue after the filters were added, hence it's closed. If no one continued commenting, we of course couldn't know the issue's state.

Anyways, if any volunteers are able to reproduce, they well re-open.

**Comment by Yuki2718** [comment]
Though I can reproduce on `https://www.ebay.com/sch/i.html?_from=R40&_nkw=ps5&_sacat=0`, it's almost "Can't fix" when they even counter JS fixes.

**Comment by Yuki2718** [comment]
I keep this opened since it might be worth @gorhill 's investigation why `ebay.com##body:xpath(//li[contains(@class, "s-item ")]//span[@aria-labelledby=ancestor::body//span[@class="clipped" and @id and not(following-sibling::span[1 and @aria-labelledby])]/@id]/ancestor::li[contains(@class, "s-item ")])` doesn't work even with `:watch-attr()` as I said in https://github.com/uBlockOrigin/uAssets/issues/23491#issuecomment-2104309049 The rule works on AG MV2 extension.

**Comment by sha-265** [comment]
> No one was able to reproduce the issue after the filters were added, hence it's closed. If no one continued commenting, we of course couldn't know the issue's state.

No said he is unable to reproduce, because no one asked to try it. I can't even find this filter in my uBO.

**Comment by gorhill** [comment]
> It might possibly be that performance-hit reached the limit,

Yes, that filter is too CPU expensive to enforce:

![image](https://github.com/user-attachments/assets/c0ff0771-83a3-4c63-98dd-5a8cf110be6d)

Execution time well over a second per DOM mutation processing cycle, so uBO disables it.

**Comment by stephenhawk8054** [comment]
> > No one was able to reproduce the issue after the filters were added, hence it's closed. If no one continued commenting, we of course couldn't know the issue's state.
> 
> No one said he is unable to reproduce, because no one asked to try it. I can't even find this filter in my uBO.

There were 2 comments for 2 filters above. Anyone who wants to test can read the above:

- https://github.com/uBlockOrigin/uAssets/issues/23491#issuecomment-2127032495
- https://github.com/uBlockOrigin/uAssets/issues/23491#issuecomment-2127525416

We are not customer service to ask each person for testing, everyone can read the comments.

---

Btw, this one looks working for me, but I'm not too sure:

```adb
ebay.com##+js(rpnt, script, (function(scope){, '(()=>{document.querySelectorAll("li.s-item").forEach((item=>{const divItem=item.querySelector("span[data-w] > div");const styles=window.getComputedStyle(divItem);let sums=0;for(const style of["padding-top","transform","margin-top"]){if(style==="transform"){const matrix=new WebKitCSSMatrix(styles.webkitTransform);sums+=matrix.m42}else{sums+=window.parseInt(styles[style].split("px")[0])}}if(sums===0){item.style.display="none"}}))})();(function(scope){')
```

**Comment by wertzuz** [comment]
On some searches for example: 
https://www.ebay.de/sch/i.html?_nkw=handy
there is a top-banner-ad shown
only verified on ebay.de
![image](https://github.com/user-attachments/assets/4b49737c-f779-486e-9c31-cf7bf316c9fe)

**Comment by Yuki2718** [comment]
> On some searches for example: https://www.ebay.de/sch/i.html?_nkw=handy there is a top-banner-ad shown only verified on ebay.de 

Looks to be self-promotion though it says `Anzeige`.

**Comment by keplx** [comment]
> Execution time well over a second per DOM mutation processing cycle, so uBO disables it.

Can each determine on his own his tolerance for waiting (processing)? then adjusting some setting per domain or globally?

**Comment by u-RraaLL** [comment]
No. uBO was built with efficiency in mind. No code changes will be made contrary to that. In fact, all prior changes that proved to lower the efficiency were fixed (where possible) or reversed (where not).

**Comment by peace2000** [comment]
Feel free to try this userscript: https://gist.githubusercontent.com/peace2000/dbbd9172c737567ad4c8c132dd770a1f/raw/f426090152b53905687625b03f85968100d052fa/gistfile1.txt

**Comment by binarynoise** [comment]
Would it be possible to delay filter execution?
I noticed that ads (in item details page) are still okish to target, but eBay seems to use JS to just place the ad somewhere else in the DOM if it finds out that the path it placed the ad is blocked.
Maybe it would work better if we waited until DOM has reached some stable state after a few seconds and apply filters only then.

**Comment by Yuki2718** [comment]
Yes, that will be useful not only for this case, but for many other anti-adblock cases too. I once suggested the idea in List Authors Chat.
@gorhill Something like `hide-on-condition` to apply cosmetic hiding only after a specific event is fired (e.g. scroll, DOMContentLoaded) and/or specified delay would have helped many anti-adblock cases in past, as most sites stop checking display state at some point. I combined `+js(ra, , , complete)` with a cosmetic filter several times for this (e.g. https://github.com/uBlockOrigin/uAssets/commit/b69cbe6b3380e05aff2cd02e62c15a7b8c19599f), which I have long been feeling inconvenient and can trivially be circumvented.

[Update] The delay should be additive to condition (e.g. 1000ms after DOMContentLoaded), and also convenient is using a selector as a condition (hiding a warning once it appeared + delay), which should have worked for `pvpoke-re.com` where `wtach-attr` was too quick to hide.

**Comment by ghost** [comment]
Shouldn't this issue be closed until maybe a fix will be available? It's been quite some time since the last comment.

**Comment by wertzuz** [comment]
on the startpage there are a few fixable self promotion ads again i think https://www.ebay.com/

https://github.com/user-attachments/assets/e6e6256b-1fbf-437f-bc3d-e8b9a74f908a

**Comment by Yuki2718** [comment]
> on the startpage there are a few fixable self promotion ads again i think https://www.ebay.com/
> 
> https://github.com/user-attachments/assets/e6e6256b-1fbf-437f-bc3d-e8b9a74f908a

Not too annoying

**Comment by mapx-** [comment]
https://github.com/uBlockOrigin/uAssets/issues/32716#issuecomment-4324440903

**Comment by wertzuz** [comment]
Isn't self promotion also a thing unlock aggressive blocks in some filterlists, since many sites can get quite annoying with it.
